api/src/main/java/org/pentaho/platform/api/genericfile/GenericFilePath.java

```diff
@@ -0,0 +1,333 @@
+   * the first segment.
+   *
+   * @return The first segment.
+   */
+  @NonNull
+  public String getFirstSegment() {
+    return segments.get( 0 );
+  }
+
+  /**
+   * Gets the path's <i>rest</i> segments.
+   * <p>
+   * The rest segments are simply the ones following the {@link #getFirstSegment() first segment}.
+   *
+   * @return The list of rest segments, possibly empty.
+   */
+  @NonNull
+  public List<String> getRestSegments() {
+    return segments.subList( 1, segments.size() );
+  }
+
+  /**
+   * Gets a value that indicates if the path has a scheme.
+   * <p>
+   * For more information, see {@link #getFirstSegment()}.
+   *
+   * @return {@code true}, if the path has a scheme; {@code false}, otherwise.
+   * @see #getScheme()
+   */
+  public boolean hasScheme() {
+    return !PATH_SEPARATOR.equals( getFirstSegment() );
+  }
+
+  /**
+   * Gets the path's scheme.
+   * <p>
+   * For more information, see {@link #getFirstSegment()}.
+   *
+   * @return The path's scheme if it has one; {@code null}, if not.
+   * @see #hasScheme()
+   */
+  @Nullable
+  public String getScheme() {
+    String providerRoot = getFirstSegment();
+    return hasScheme()
+      ? providerRoot.substring( 0, providerRoot.length() - SCHEME_SUFFIX.length() )
+      : null;
+  }
+
+  /**
+   * Gets the path's segments.
+   *
+   * @return An immutable list of path segments.
+   * @see #getFirstSegment()
+   * @see #getRestSegments()
+   */
+  @NonNull
+  public List<String> getSegments() {
+    return segments;
+  }
+
+  /**
+   * Gets the path's normalized string representation.
+   *
+   * @return The string representation.
+   */
+  @Override
+  public String toString() {
+    return path;
+  }
+
+  /**
+   * Gets the parent generic path instance.
+   * <p>
+   * The parent path of <i>provider root paths</i> is {@code null}.
+   *
+   * @return The parent generic path instance, if any; {@link null}, if none.
+   */
+  @Nullable
+  public GenericFilePath getParent() {
+    if ( segments.size() == 1 ) {
+      return null;
+    }
+
+    return new GenericFilePath( segments.subList( 0, segments.size() - 1 ) );
+  }
+
+  /**
+   * Parses path given its string representation, while ensuring a path can be returned.
+   * <p>
+   * If a {@code null} or blank string is specified, then an exception is thrown.
+   * <p>
+   * Otherwise, this method delegates to {@link #parse(String)}.
+   * @param path The path string to parse.
+   * @return The generic path instance.
+   * @throws InvalidPathException If the path is {@code null}, blank, or otherwise invalid. Specifically, if the path's
+   *                              root segment is not either a {@link #PATH_SEPARATOR} or a scheme followed by the
+   *                              {@code ://} suffix.
+   */
+  @NonNull
+  public static GenericFilePath parseRequired( @Nullable String path ) throws InvalidPathException {
```

Comment:
not a blocker. Why was the decision made to move this one check out of parse method below?

Reply:
looks like there are a couple cases where you don't want to check for null returned path, so you created this layer for cases where you do want to check for it.

Maybe call the method parseSafely or strictParse or something about the fact that this has a check in it whereas parse doesn't. To me, "required" does not convey this distinction.

Reply:
validatedParse?

Reply:
I mean there are checks/validations in parse, so... hmmm maybe the non-nullness?

parseNonNull or nonNullParse? something like that?

Reply:
This is a suffix I use for these situations. Something like `getRequired(...)`, where you also have a method such as `get(..)` which is lenient and returns null. I read this as "parse value which is expected to be required".

When methods are written the other way around, I sometimes use the the "Lenient", or "Optional" suffix: `getOptional(...)` vs `get(...)`.

I see this method as essentially syntax sugar. Both accept `@Nullable` arguments, but only `parseRequired(..)` ensures `@NonNull` return value, because it throws an `InvalidPathException` otherwise.

I am not sure your suggestions would be clearer. I'd maybe prefer switching the names to `parseOptional()` and `parse()`, if you think it would be more easily understood this way...

My comments on your suggestions:
- `parseSafely`: both are safes regarding some aspects of validation
- `validatedParse`: the normal parse also validates the structure of the path, except it returns null for the special case of a completely empty path.
- `parseStrict`: being strict would apply to other aspects of validation as well, which would be seen more strictly. But there's none at the moment. And not being filled-in is always a special kind of "invalid".
- `parseNonNull`: in a way, this may give the impression that the method does not accept null values, yet it does. Also, the "required" validation is also including the special "empty" and "blank" cases. Maybe, `parseNonBlank`.